docs/dev-guides/java-style.md

```diff
@@ -0,0 +1,166 @@
+
+### Java 8 features
+
+Use Java 8 features where helpful. See [https://leanpub.com/whatsnewinjava8/read](https://leanpub.com/whatsnewinjava8/read) and [https://www.oracle.com/java/technologies/javase/8-whats-new.html](https://www.oracle.com/java/technologies/javase/8-whats-new.html) for more details.
+
+Here is a quick run-down of Java 8 topics:
+
+*   **java.time.* – **the JavaTime library _should_ be used for date-time manipulation. Duration is especially common. For non-core extensions, consider the [https://www.threeten.org/threeten-extra/](https://www.threeten.org/threeten-extra/) library.
+*   **Lambda expressions** – _prefer_ lambda expressions for short and simple bits of code. If a lambda gets too complex, extract it into a named method.
+*   **Method references** – _prefer_ method references where possible (e.g. `someStream().map(MyClass::getBar)`).
+*   **Streams** – _prefer_ using Streams for operations that can be framed as a simple, compact stream expression. On the other hand: streams can interfere with exception handling, and complex streams may be more difficult to reason about than equivalent non-stream code, so judgement is required.
+
+### Javadoc
+
+Javadoc comments (`/** … */`) are meant to explain the intended purpose of the element they are attached to. See the [style guide](https://google.github.io/styleguide/javaguide.html) for a canonical list of when Javadoc is required.
+
+Some guidelines for writing good Javadoc:
+
+*   Spend more time documenting classes that will be frequently used or updated. Documentation mitigates against the risk of future misinterpretation.
+*   Block tags (`@param` and `@return`) are generally considered _optional_:
+    *   Use `@param` block tags when detail is required to clearly convey the meaning of each function parameter.
+    *   Use a `@return` block tag if the main text cannot clearly express the meaning of the returned value in a natural way.
+
+### Methods
+
+Some guidelines around Java methods:
+
+*   Keep methods short and simple.
+*   Avoid unwieldy parameter lists.
+*   Beware of accepting or returning null.
+*   _Prefer_ general types for parameters and specific types for return values.
+*   Avoid output parameters.
+
+See [EJ3 #49-56](https://github.com/david-sauvage/effective-java-summary#methods) for more tips.
+
+### Null
+
+Avoid nullable references where possible! Of course, this isn’t always possible, but null checks and `@Nullable` annotations should always be viewed with healthy skepticism.
```

Comment:
One place I like to see `@Nullable` is to flag that a utility method is doing the null check for me. If we're not going to have optional parameters, the only way to avoid nullable parameters is with overloading, which only gets you so far. Builder FTW.

Reply:
I agree – and I think that's in line with the guidance in this section. Happy to adjust language if there is any clear mis-guidance.

Reply:
Maybe "If a function's contract deliberately accepts null values for a parameter (for example, so that it can handle the situation), then @Nullable may be appropriate."

I also tend to use `@Nullable` on private class members that are exposed as `Optional` methods.